src/librustc_typeck/check/_match.rs

```diff
@@ -112,14 +115,74 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 self.check_expr_with_expectation(&arm.body, expected)
             };
             all_arms_diverge &= self.diverges.get();
+
+            // When we have a `match` as a tail expression in a `fn` with a returned `impl Trait`
+            // we check if the different arms would work with boxed trait objects instead and
+            // provide a structured suggestion in that case.
+            let suggest_box = match (
+                orig_expected,
+                self.body_id
+                    .owner
+                    .to_def_id()
+                    .as_local()
+                    .and_then(|id| self.ret_coercion_impl_trait.map(|ty| (id, ty))),
+            ) {
+                (Expectation::ExpectHasType(expected), Some((id, ty)))
```

Comment:
```suggestion
                (Expectation::ExpectHasType(expected), Some((body_id, ty)))
```
unclear what `id` means here